janus_server/src/aggregator.rs

```diff
@@ -933,6 +1010,171 @@ impl VdafOps {
+    async fn handle_aggregate_share(
+        &self,
+        datastore: &Datastore,
+        task: &Task,
+        aggregate_share_req: &AggregateShareReq,
+    ) -> Result<AggregateShareResp, Error> {
+        match self {
+            VdafOps::Prio3Aes128Count(_, _) => {
+                Self::handle_aggregate_share_generic::<
+                    <Prio3Aes128Count as Vdaf>::AggregateShare,
+                    FieldError,
+                >(datastore, task, aggregate_share_req)
+                .await
+            }
+            VdafOps::Prio3Aes128Sum(_, _) => {
+                Self::handle_aggregate_share_generic::<
+                    <Prio3Aes128Sum as Vdaf>::AggregateShare,
+                    FieldError,
+                >(datastore, task, aggregate_share_req)
+                .await
+            }
+            VdafOps::Prio3Aes128Histogram(_, _) => {
+                Self::handle_aggregate_share_generic::<
+                    <Prio3Aes128Histogram as Vdaf>::AggregateShare,
+                    FieldError,
+                >(datastore, task, aggregate_share_req)
+                .await
+            }
+
+            #[cfg(test)]
+            VdafOps::Fake(_) => {
+                Self::handle_aggregate_share_generic::<
+                    <fake::Vdaf as Vdaf>::AggregateShare,
+                    Infallible,
+                >(datastore, task, aggregate_share_req)
+                .await
+            }
+        }
+    }
+
+    async fn handle_aggregate_share_generic<A, E>(
+        datastore: &Datastore,
+        task: &Task,
+        aggregate_share_req: &AggregateShareReq,
+    ) -> Result<AggregateShareResp, Error>
+    where
+        A: Aggregatable + Send + Sync,
+        E: std::fmt::Debug,
+        Vec<u8>: for<'a> From<&'a A>,
+        for<'a> A: TryFrom<&'a [u8], Error = E>,
+    {
+        // TODO(timg): What should we do if any of the aggregation jobs handling reports relevant to
+        // the aggregate share request aren't finished yet? Helper could return an error like
+        // "EAGAIN" to leader, or could trust leader not to issue an AggregateShareReq until all of
+        // leader's aggregate jobs are done (which should imply that helper's are done).
+
+        let total_aggregate_share = datastore
+            .run_tx(move |tx| {
+                let task = task.clone();
+                let aggregate_share_req = aggregate_share_req.clone();
+                Box::pin(async move {
+                    let batch_aggregations = tx
+                        .get_batch_aggregations::<A, E>(task.id, aggregate_share_req.batch_interval)
+                        .await?;
+
+                    let mut total_report_count = 0;
+                    let mut total_checksum = [0u8; 32];
+                    let mut total_aggregate_share: Option<A> = None;
+
+                    for batch_aggregation in &batch_aggregations {
+                        // TODO(timg): This satisfies the max lifetime requirement in §4.6 but it
```

Comment:
+1.

If we end up including the aggregation parameter in the `batch_aggregations` table, I think we could do this by counting the number of `batch_aggregations` there are for the `(batch window, aggregation parameter)`.

Reply:
Correctly enforcing max_batch_lifetime requires more persistence on the helper side, so I punted for now to avoid bloating this PR.